File: tetracorder5.27/i4pack4.r
      function i4pack4(ibuff,npix,numsiz)
      integer*4 i4pack4
#
#  this function packs 32 bits/pixel files to 8 or 16 bits/pixel
#
#  Nov. 2008   K. Eric Livo
#  16-bit fixed;  Jan. 21, 2011 - Livo
#
#     ibuff is 32bit (I*4) data
#     npix = number of pixels
#     numsiz = 8 or 16 (bit)
#
#     isys = system; 0=IEEE (MSB), 1=INTEL (LSB)
#
#     8bit, IEEE:isys1=0;  INTEL:isys1=3
#     16 bit, IEEE:isys1=-1, isys2=0
#             INTEL:isys1=-3, isys2=-2
#
      logical*1 ibuff(1)
      integer*4 npix, numsiz
      integer*4 i, j, isys, isys1, isys2, n16pix

#  set system to IEEE or INTEL
#HPUX      isys = 0
#IA64HPUX      isys = 0
#LINUX      isys = 1

#  8 bit packs
      if (numsiz == 8) {
         isys1 = isys*3

         do i=1, npix {
            ibuff(i)=ibuff(i*4 - isys1)
         }
      }

#  16 bit packs
      if (numsiz == 16) {
         isys1 = 1 +(isys * 2)
         isys2 = (isys * 2)
         i=1

         do j=1, npix {
            ibuff(i)=ibuff(j*4 - isys1)
            ibuff(i+1)=ibuff(j*4 - isys2)
            i=i+2
         }
      }

      i4pack4 = 0
      return
      end

File: tetracorder6.00/i4pack4.c
/*__________________________________________________________________
 |  TITLE:                   PIXEL PACKER (Version 4)              |
 |  PROGRAMMER:              Barry J. Middlebrook                  |
 |-----------------------------------------------------------------|
 |  DESCRIPTION:                                                   |
 |              This subroutine is for packing pixels into 32 bit  |
 |              numbers.  It handles 8, 16 and 32 bit numbers.     |
 |                                                                 |
 |-----------------------------------------------------------------|
 |  VARIABLES:                                                     |
 |            ibuff	- incoming array containing pixels to be   |
 |                      packed (32 bit -> 8 or 16 bit)             |
 |            npix	- number of pixels to pack                 |
 |            numsiz	- number of bits to pack to (32->numsiz)   |
 |            pak	- register variable the pixels are packed  |
 |                      in                                         |
 |            maxoff	- maximum offset from array starting       |
 |                      address                                    |
 |            inc	- increment to move through input array    |
 |            ercode	- return error code (1 - error, 0 - ok)    |
 |                                                                 |
 |_________________________________________________________________|
*/

union iobuff
{
	unsigned char		chbuff[8192];
	unsigned short int	i2buff[4096];
	unsigned long int	i4buff[2048];
};

/*HPUX
long int i4pack4 (ibuff,npix,numsiz)
HPUX*/

/*NONHPUX
long int i4pack4_ (ibuff,npix,numsiz)
NONHPUX*/

/*IA64HPUX
long int i4pack4_ (ibuff,npix,numsiz)
IA64HPUX*/

/*LINUX
long int i4pack4_ (ibuff,npix,numsiz)
LINUX*/

union iobuff	*ibuff;
long int	*npix,*numsiz;

{
	register long int	i,pak,j;
	long int		maxoff,inc,ercode;

/* Initialize variables */
	maxoff = (*npix << 2);  /* Equivalent to multiplying by 4 */
	inc = 128/(*numsiz);

/* Select between 8 (case 16) and 16 (case 8) bit numbers */
	switch (inc)
	{
	   case 16:	/* Pack 32-bit pixels to 8-bit numbers */	
			for (i = j = 0; i < maxoff; i += inc, j++) 
			{
				pak = ibuff -> chbuff[i+3];
				pak = (pak << 8) | ibuff -> chbuff[i+7];
				pak = (pak << 8) | ibuff -> chbuff[i+11];
				pak = (pak << 8) | ibuff -> chbuff[i+15];

				ibuff -> i4buff[j] = pak;
			}
			return(0);

	   case 8:	/* Pack 32-bit pixels to 16-bit numbers */	
			for (i = j = 0; i < maxoff; i += inc, j++) 
			{
				pak = ibuff -> chbuff[i+2];
				pak = (pak << 8) | ibuff -> chbuff[i+3];
				pak = (pak << 8) | ibuff -> chbuff[i+6];
				pak = (pak << 8) | ibuff -> chbuff[i+7];

				ibuff -> i4buff[j] = pak;
			}
			return(0);

	   default:	return (1);

	}
}
